### DISCUSSION TITLE
how to know whether `AppState` is changed using `onPointerUp` API

### DISCUSSION THREAD

**Discussion by Lslightly** [excalidraw/excalidraw#8959] [Q&A]
I have an idea to record the state, like exporting png, of excalidraw as `undo` and `redo` histories do. After capturing a series of png, splice them all to get a video.

[`onChange`](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api/props/excalidraw-api#onchange) does not meet my request because moving a component will trigger the callback, which is unexpected. I just want to record the result of a moving action instead of details in the moving process. So I find [`onPointerUp`](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api/props/excalidraw-api#onpointerup). When a pointerup event is triggered, it's quite likely that an action has been completed. In such a case, I can record the state by exporting it as a png or in some other format. However, since there's only a certain probability that an action will actually be executed, I need to determine whether the state has truly changed.

I want to judge whether `AppState` is changed when the callback function of [`onPointerUp`](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api/props/excalidraw-api#onpointerup) is called. But I'm overwhelmed by too many APIs. How to judge whether the `AppState` is changed? Thanks a lot.

> `AppStateChange.calculate` might be a good choice but it seems that I need to store old `AppState`, which might consume large memory and not so efficient I think.

I find [the comment](https://github.com/excalidraw/excalidraw/discussions/8721#discussioncomment-11086139) helpful. But [excalidraw-animate](https://github.com/dai-shi/excalidraw-animate) is for animation, but not for recording I think.

**Comment by Lslightly** [comment]
Is it possible to register listeners in [`onStoreIncrementEmitter`](https://github.com/excalidraw/excalidraw/blob/798c795405a8f5858ffb5fef473730aa08afea59/packages/excalidraw/store.ts#L199) such that after finishing a complete update/modify action, like moving a component from one place to another place/change the color of a component, some additional actions can be taken.

**Reply by Mrazator** [reply]
Technically you might use it, though this API is not exposed yet. There will likely be `onIncrement` handler as part of `props` & `excalidrawApi` in the future. We might also add to the API not just the increments (~props of elements that have changed), but also the whole changed elements (vice versa for appstate).

**Comment by Lslightly** [comment]
I found a tool [excalidraw-claymate](https://github.com/dai-shi/excalidraw-claymate) but it's not good enough. I like Rust's default move semantic of objects but the current implementation of the tool has clone semantic. 

Besides, I found an [use case](https://www.reddit.com/r/ObsidianMD/comments/1b5plue/can_we_export_slides_to_powerpoint/) in reddit, which exports images to PowerPoint and adds animations.
